Deploying_Seq2Seq_Model_with_Hybrid_Frontend.ipynb

```diff
@@ -0,0 +1,720 @@
+    "    s = re.sub(r\"[^a-zA-Z.!?]+\", r\" \", s)\n",
+    "    return s\n",
+    "\n",
+    "\n",
+    "# Takes string sentence, returns sentence of word indexes\n",
+    "def indexesFromSentence(voc, sentence):\n",
+    "    return [voc.word2index[word] for word in sentence.split(' ')] + [EOS_token]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define Evaluation\n",
+    "\n",
+    "### Greedy Search Decoder\n",
+    "\n",
+    "As in the chatbot tutorial, we use a `GreedySearchDecoder` module to facilitate the actual decoding process. This module has the trained encoder and decoder models as attributes, and drives the process of encoding an input sentence (a vector of word indexes), and iteratively decoding an output response sequence one word (word index) at a time. \n",
+    "\n",
+    "Encoding the input sequence is straightforward: simply forward the entire sequence tensor and its corresponding lengths vector to the `encoder`. It is important to note that this module only deals with one input sequence at a time, **NOT** batches of sequences. Therefore, when the constant **1** is used for declaring tensor sizes, this corresponds to a batch size of 1. To decode a given decoder output, we must iteratively run forward passes through our decoder model, which outputs softmax scores corresponding to the probability of each word being the correct next word in the decoded sequence. We initialize the `decoder_input` to a tensor containing an *SOS_token*. After each pass through the `decoder`, we *greedily* append the word with the highest softmax probability to the `decoded_words` list. We also use this word as the `decoder_input` for the next iteration. The decoding process terminates either if the `decoded_words` list has reached a length of *MAX_LENGTH* or if the predicted word is the *EOS_token*.\n",
+    "\n",
+    "### Hybrid Frontend Notes:\n",
+    "\n",
+    "The `forward` method of this module involves iterating over the range of $[0, max\\_length)$ when decoding an output sequence one word at a time. Because of this, we should use **scripting** to convert this module to Torch Script. Unlike with our encoder and decoder models, which we can trace, we must make some necessary changes to the `GreedySearchDecoder` module in order to initialize an object without error. In other words, we must ensure that our module adheres to the rules of the scripting mechanism, and does not utilize any language features outside of the subset of Python that Torch Script includes.\n",
+    "\n",
+    "To get an idea of some manipulations that may be required, we will go over the diffs between the `GreedySearchDecoder` implementation from the chatbot tutorial and the implementation that we use in the cell below. Note that the lines highlighted in red are lines removed from the original implementation and the lines highlighted in green are new.\n",
+    "\n",
+    "![diff](images/gsd_diff.png)\n",
```

Comment:
I don't like this diff, since using `device` and `dtype` as attributes on the tensor constructors (e.g. `torch.ones`) likely should have been done in the original code to prevent cross-device copies and unnecessary casts